src/CorrelatedNoiseDetection.cc

```diff
@@ -0,0 +1,192 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST/AURA.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "lsst/pex/exceptions.h"
+#include "lsst/afw/math/LeastSquares.h"
+#include "lsst/meas/algorithms/CorrelatedNoiseDetection.h"
+
+namespace lsst { namespace meas { namespace algorithms {
+
+
+afw::image::Image<float> measureCorrelationKernel(
+    afw::image::MaskedImage<float> const & mi,
+    int radius,
+    afw::image::MaskPixel badBitMask
+) {
+    afw::image::Image<float> result(2*radius + 1, 2*radius + 1);
+    afw::image::Image<int> count(result.getDimensions());
+    int const width = mi.getWidth();
+    int const height = mi.getHeight();
+    // iterate over pixels in the MaskedImage, skipping any that meet our bad mask criteria
+    for (int y1 = 0; y1 < height; ++y1) {
+        int const y2a = std::max(0, y1 - radius);
+        int const y2b = std::min(height, y1 + radius + 1);
+        for (int x1 = 0; x1 < width; ++x1) {
+            if ((*mi.getMask())(x1, y1) & badBitMask) {
+                continue;
+            }
+            float z = (*mi.getImage())(x1, y1) / (*mi.getVariance())(x1, y1);
+            // iterate over neighboring pixels, with bounds set to avoid image boundaries
+            int const x2a = std::max(0, x1 - radius);
+            int const x2b = std::min(height, x1 + radius + 1);
+            for (int y2 = y2a; y2 < y2b; ++y2) {
+                auto miIter = mi.row_begin(y2) + x2a;
+                auto outIter = result.row_begin(radius + y2 - y1) + radius + x2a - x1;
+                auto countIter = count.row_begin(radius + y2 - y1) + radius + x2a - x1;
+                for (int x2 = x2a; x2 < x2b; ++x2, ++miIter, ++outIter, ++countIter) {
+                    if (miIter.mask() & badBitMask) {
+                        continue;
+                    }
+                    *outIter += z*miIter.image();
```

Comment:
Why are you dividing the variance value from only the first pixel?  Shouldn't be something like ```sqrt(var(x1,y1))*sqrt(var(x2,y2))```?

Reply:
I think you're right - I was originally thinking of this as part of the stationarity assumption, but I don't need to assume the variance is stationary to assume that the correlations are.